### ISSUE TITLE
[c++20][Modules] incorrect error regarding ambiguous specialization

### ISSUE THREAD

**Issue by mordante** [llvm/llvm-project#62943]
Tested with a recent Clang nightly build, using libc++'s `std` module.

https://github.com/llvm/llvm-project/blob/11ea40cff5413057d823a4b3ac5ac419b674dc56/libcxx/include/__iterator/readable_traits.h#L84
```hpp
template <class _Ip>
using iter_value_t = typename conditional_t<__is_primary_template<iterator_traits<remove_cvref_t<_Ip> > >::value,
                                            indirectly_readable_traits<remove_cvref_t<_Ip> >,
                                            iterator_traits<remove_cvref_t<_Ip> > >::value_type;
```

Here the `indirectly_readable_traits` causes an ambiguous specialization error in the following code:
```cpp
import std;

template <class It>
class contiguous_iterator {
  It it_;
public:
  typedef std::contiguous_iterator_tag iterator_category;
  typedef typename std::iterator_traits<It>::value_type value_type;
  typedef typename std::iterator_traits<It>::difference_type difference_type;
  typedef It pointer;
  typedef typename std::iterator_traits<It>::reference reference;
  typedef typename std::remove_pointer<It>::type element_type;

  constexpr It base() const { return it_; }

  constexpr contiguous_iterator() : it_() {}
  constexpr explicit contiguous_iterator(It it) : it_(it) {}

  constexpr reference operator*() const { return *it_; }
  constexpr pointer operator->() const { return it_; }
  constexpr reference operator[](difference_type n) const { return it_[n]; }

  constexpr contiguous_iterator& operator++() {
    ++it_;
    return *this;
  }
  constexpr contiguous_iterator& operator--() {
    --it_;
    return *this;
  }
  constexpr contiguous_iterator operator++(int) { return contiguous_iterator(it_++); }
  constexpr contiguous_iterator operator--(int) { return contiguous_iterator(it_--); }

  constexpr contiguous_iterator& operator+=(difference_type n) {
    it_ += n;
    return *this;
  }
  constexpr contiguous_iterator& operator-=(difference_type n) {
    it_ -= n;
    return *this;
  }
  friend constexpr contiguous_iterator operator+(contiguous_iterator x, difference_type n) {
    x += n;
    return x;
  }
  friend constexpr contiguous_iterator operator+(difference_type n, contiguous_iterator x) {
    x += n;
    return x;
  }
  friend constexpr contiguous_iterator operator-(contiguous_iterator x, difference_type n) {
    x -= n;
    return x;
  }
  friend constexpr difference_type operator-(contiguous_iterator x, contiguous_iterator y) { return x.it_ - y.it_; }

  friend constexpr bool operator==(const contiguous_iterator& x, const contiguous_iterator& y) {
    return x.it_ == y.it_;
  }
  friend constexpr bool operator!=(const contiguous_iterator& x, const contiguous_iterator& y) {
    return x.it_ != y.it_;
  }
  friend constexpr bool operator<(const contiguous_iterator& x, const contiguous_iterator& y) { return x.it_ < y.it_; }
  friend constexpr bool operator<=(const contiguous_iterator& x, const contiguous_iterator& y) {
    return x.it_ <= y.it_;
  }
  friend constexpr bool operator>(const contiguous_iterator& x, const contiguous_iterator& y) { return x.it_ > y.it_; }
  friend constexpr bool operator>=(const contiguous_iterator& x, const contiguous_iterator& y) {
    return x.it_ >= y.it_;
  }
};

using ContiguousIter = contiguous_iterator<const int*>;

// enabling this "fixes" the issue.
//static_assert(std::same_as<std::indirectly_readable_traits<float*>::value_type, float>);
static_assert(std::contiguous_iterator<ContiguousIter>);
```
This code is reduced from the `ranges::data` test. The interesting part is when the `indirectly_readable_traits` is used in a `static_assert` for an unrelated type, then the code compiles.  When using this code without modules it works too.

**Comment by ChuanqiXu9** [comment]
Given it is a little little bit hard to play around std modules now, I'd like to investigate the issue several weeks later.

**Comment by mordante** [comment]
That sounds sensible to me.

**Comment by ChuanqiXu9** [comment]
> I've create a patch for module support https://reviews.llvm.org/D151814 you could test with that patch or wait until it lands.

I'd like to wait for it lands. Since I am looking at support modules in clangd recently: https://github.com/clangd/clangd/issues/1293. I feel it may be less overhead to test this after it lands.

**Comment by ChuanqiXu9** [comment]
I took a simple look and find the compiler complains about the ambiguous partial specialization of indirectly_readable_traits: https://github.com/llvm/llvm-project/blob/18eb31c096167a1f6185f05338494c82a509ff9e/libcxx/include/__iterator/readable_traits.h#L58-L75

It looks valid to me at the first sight since the reproducer type has both value_type and element_type.

(I don't know why it is OK to compile without modules yet)

**Comment by ChuanqiXu9** [comment]
I got the problem. It is much more complex than I thought. I sent a mail to WG21. If you're interested, you can take a look at EWG/CWG's mailing list.

**Comment by mordante** [comment]
Thanks for the notification! I'm not subscribed to these lists, but I'll look in the archives.
